Merge feature/use-dllimport-generator to main

jkoritzinsky opened the pull request:
Merge the DllImportGenerator experiment to the main branch for .NET 7.

This should be merged with "Create a merge commit"

ASP.NET JSON macro-benchmark: https://github.com/dotnet/runtime/pull/59579#issuecomment-938878840
Libraries micro-benchmarks: https://github.com/dotnet/runtime/pull/59579#issuecomment-940286862

Items we will address once the initial merge happens: https://github.com/dotnet/runtime/issues/60212

cc @stephentoub

AaronRobinsonMSFT committed:
Initial scaffolding for DllImportGenerator
  - Includes identification of GeneratedDllImportAttribute
  - Generation of enclosing types for partial methods
  - Basic tests for compilation of generation code


Commit migrated from https://github.com/dotnet/runtimelab/commit/56076d30fd1e83f441d316094a69577b54b8799b

AaronRobinsonMSFT committed:
Add GENERATE_FORWARDER define to demonstrate the ability to forward
  calls directly to the real DllImport from the generated one.

Add test for handling RefKind (i.e. in, ref, out).


Commit migrated from https://github.com/dotnet/runtimelab/commit/36706439bd965d05baffd52c579f2a375369bb54

AaronRobinsonMSFT committed:
Check for code snippet compilation errors
 - User defined source needs to be "correct"


Commit migrated from https://github.com/dotnet/runtimelab/commit/f9f705cc62acc4d1b9a3cd10273bea70b1128e71

AaronRobinsonMSFT committed:
Add example that runs on Windows to the demo project.

It is interesting to step into the generated method and observe the UX.


Commit migrated from https://github.com/dotnet/runtimelab/commit/35c9adc2b099a51aff24992e15380a1f26a60d39

AaronRobinsonMSFT committed:
Update readme with existing experiments (dotnet/runtimelab#8)

* Update readme with existing experiments
* Add "CreateAnExperiment.md"

Commit migrated from https://github.com/dotnet/runtimelab/commit/542dd851129f974632dddaa24a2b9cb214fd5975

jkotas committed:
Update CreateAndExperiment.md (dotnet/runtimelab#32)



Commit migrated from https://github.com/dotnet/runtimelab/commit/4072ab96916749f598fdab82d3a83c66b9cdbffb

AaronRobinsonMSFT committed:
Add testing for attribute discovery in cases where the user prefixes
the attribute name.

Add test cases for compilation failures.


Commit migrated from https://github.com/dotnet/runtimelab/commit/d2f6f5a752a88349c80a37e34e55357c0309628b

AaronRobinsonMSFT committed:
Merge pull request dotnet/runtimelab#35 from AaronRobinsonMSFT/failure_test_case

Failure test cases.

Commit migrated from https://github.com/dotnet/runtimelab/commit/48b836175514de82f3b8b329c4abdbc0ade20ca6

AaronRobinsonMSFT committed:
Collect DllImport details through Roslyn semantic model (dotnet/runtimelab#36)

* Collect DllImport details through Roslyn semantic model

Create an assembly that provides the GeneratedDllImportAttribute
  which allows use of the semantic model.

* Collect MarshalAsAttribute data

* Collect MarshalAsAttribute data
Convert P/Invoke centric data structures to handle positional
  type information for managed/unmanaged scenarios.

Commit migrated from https://github.com/dotnet/runtimelab/commit/21e634efd416d3457d171d0db6efe505b48c711d

AaronRobinsonMSFT committed:
Special case boolean/char types in managed/native signatures. (dotnet/runtimelab#47)



Commit migrated from https://github.com/dotnet/runtimelab/commit/39ce2af428817677201f0f2fdd79c2a3bfddcefe

jkoritzinsky committed:
Add struct marshalling design doc. (dotnet/runtimelab#45)

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

Commit migrated from https://github.com/dotnet/runtimelab/commit/e50c8c26eaef29e47a455c0781eda10a3786957e

jkoritzinsky committed:
Add the struct marshalling attributes to Ancillary.Interop and add an analyzer that validates manual usage. (dotnet/runtimelab#61)

Co-authored-by: Elinor Fung <elfung@microsoft.com>

Commit migrated from https://github.com/dotnet/runtimelab/commit/05d6ef236f7bd698870b17e8abb1b7c6bb130198

jkoritzinsky committed:
Implement infra to support marshalling blittable generics (equivalent to .NET 5 support) (dotnet/runtimelab#80)



Commit migrated from https://github.com/dotnet/runtimelab/commit/91b131e8d9340e8591d258793dd142e7c2e7a011

safern committed:
Update docs to include convention (dotnet/runtimelab#93)



Commit migrated from https://github.com/dotnet/runtimelab/commit/4eaa5d8c47ed31d59d0d806b14ce532a569dd4de

elinor-fung committed:
Basic structure for composing stub code (dotnet/runtimelab#87)



Commit migrated from https://github.com/dotnet/runtimelab/commit/e787027eafb2c40607fc46db586b7c0071c3d8ab

jkoritzinsky committed:
Use declaring syntax as a baseline for generated syntax for containing types. (dotnet/runtimelab#108)



Commit migrated from https://github.com/dotnet/runtimelab/commit/4bb5cb0590896a2b88fd12d041d4268233220ab8

elinor-fung committed:
Move to 3.8.0-3.final Roslyn packages (dotnet/runtimelab#119)



Commit migrated from https://github.com/dotnet/runtimelab/commit/f56021b8dec78527297f97a39fa0375f6dea6619

jkoritzinsky committed:
Add validation that the StackBufferSize member exists when a stackalloc constructor exists. (dotnet/runtimelab#125)



Commit migrated from https://github.com/dotnet/runtimelab/commit/9e7e2219120dcce1bf7c3356269f743cfe1d1f31

AaronRobinsonMSFT committed:
Consume DNNE for testing (dotnet/runtimelab#124)

* Use DNNE to create native testing assets

Instead of relying on a native project use DNNE to generate
native exports for managed binaries. This allows us to avoid
writing native code for testing code generation.

* Add *.binlog to .gitignore

* Force 'UseAppHost' to fix scenario on macOS.

* Consume support in DNNE for adding native binary to project.


Commit migrated from https://github.com/dotnet/runtimelab/commit/415c6ca78b4d38409e0d569804a504ae2b8f5dad

jkoritzinsky committed:
Implement marshaler selection based on metadata model (dotnet/runtimelab#109)



Commit migrated from https://github.com/dotnet/runtimelab/commit/ebf31e6d43e96cacf3aea98738e71663ce468099

jkoritzinsky committed:
Fix true and false for the Bool marshaller (dotnet/runtimelab#130)



Commit migrated from https://github.com/dotnet/runtimelab/commit/37705c3bd1310e59ce518bd89ce32d2a99e79d0e

jkotas committed:
Update CreateAnExperiment.md

Commit migrated from https://github.com/dotnet/runtimelab/commit/8497669c8d619e3164e36ab00df221e3cd8a41b0

jkoritzinsky committed:
SafeHandle marshalling (dotnet/runtimelab#133)

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

Commit migrated from https://github.com/dotnet/runtimelab/commit/2cde5aa726f1b0ad08fd774d0bf84bd9eb5f7e62

AaronRobinsonMSFT committed:
Create test projects (dotnet/runtimelab#149)

* Rename DllImportGenerator.Test to DllImportGenerator.UnitTests

* Create an integration tests project.
Update the SLN file.
Remove SafeHandle test from Demo project.

* Update readme.md
Make the Demo project a user friendly example for learning about the methodology.

Commit migrated from https://github.com/dotnet/runtimelab/commit/778ecfa4e1b31962318ceed823882759e1f5bf4a

safern committed:
Create branch template for standalone templates.


Commit migrated from https://github.com/dotnet/runtimelab/commit/891d530a4b46bafd3fbb1146ce4e8553dbce36e3

AaronRobinsonMSFT committed:
Boolean and Delegate tests (dotnet/runtimelab#152)

* Boolean and Delegate tests

Updated to DNNE 1.0.12 with support for C# function pointers.
Renamed CBoolMarshaller to ByteBoolMarshaller.
Added integration tests for all bool marshaller types.
Added integration tests for delegates.
Style nits.

Commit migrated from https://github.com/dotnet/runtimelab/commit/6c980dff7af857ebc68518150c08a96a95a04d1d

elinor-fung committed:
Report diagnostics in generator (dotnet/runtimelab#158)



Commit migrated from https://github.com/dotnet/runtimelab/commit/98f55be7fed518bb93391d0d24341b4b5e7a56fd

AaronRobinsonMSFT committed:
Add Char marshaller (dotnet/runtimelab#212)

* Support for char marshaller.

Avoid supporting the built-in semantics for ANSI marshalling of char.
Propagate the CharSet down to the marshallers.
Convert the MarshallingGenerators.TryCreate() to a MarshallingGenerators.Create() function throwing an exception.

* Add a compatibility document.

* Add testing for char marshalling.

Commit migrated from https://github.com/dotnet/runtimelab/commit/be3f4e8b23c3740e83b3a6b20dda7ad41a077118

AaronRobinsonMSFT committed:
Fix warnings for analyzer release (dotnet/runtimelab#227)

* Fix some warnings about analyzer authoring

Commit migrated from https://github.com/dotnet/runtimelab/commit/c89fc418755eadc8dc457a298a032caa45273b8c

jkoritzinsky committed:
Enable nullability checks across DllImportGenerator (dotnet/runtimelab#213)



Commit migrated from https://github.com/dotnet/runtimelab/commit/7b2591d6531af56420092d0ae136911ae507e218

elinor-fung committed:
String marshallers for UTF-16 and UTF-8 (dotnet/runtimelab#249)



Commit migrated from https://github.com/dotnet/runtimelab/commit/ac6762f9ef20577d4847b6a55de4df7f84f2f74e

elinor-fung committed:
Disable nullable checks in integration tests (dotnet/runtimelab#253)

Fix nullability warnings in generator

Commit migrated from https://github.com/dotnet/runtimelab/commit/7bb0f191c7659b270e5f7d792922193adde7d0cc

AaronRobinsonMSFT committed:
Remove support for the custom marshaller on MarshalAsAttribute. (dotnet/runtimelab#255)

* Remove support for the custom marshaller on MarshalAsAttribute.

* Update compat document.

Commit migrated from https://github.com/dotnet/runtimelab/commit/5a0e1d94fb123008623c8ef9c26513ff57d3481e

AaronRobinsonMSFT committed:
Add initial blittable struct test (dotnet/runtimelab#230)

Create SharedTypes project to shared between NativeExports
  and integration tests.
Update DNNE to 1.0.13 for custom native argument signature support.

Commit migrated from https://github.com/dotnet/runtimelab/commit/acff10602714c8133275c17de9791304dd6cf740

safern committed:
Add standalone template for experiments that don't need runtime features (dotnet/runtimelab#238)

* Add arcade common files

* Add standalone library template for experiments that don't need runtime features

* Add documentation to README.md

* PR Feedback.

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

* Quick fixups in README.md

* PR Feedback

* Simplify yml templates

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

Commit migrated from https://github.com/dotnet/runtimelab/commit/0809d8118490cdc32b8b89c43479697a76c22e66

AaronRobinsonMSFT committed:
Factor out some nullability uses (dotnet/runtimelab#259)

* Factor out some nullability uses

Commit migrated from https://github.com/dotnet/runtimelab/commit/915905c549e8cd74e22c9e880b2de2718b04d9e9

AaronRobinsonMSFT committed:
Add event for marking the Start/Stop of source generation (dotnet/runtimelab#258)

* Add event for marking the Start/Stop of source generation

Commit migrated from https://github.com/dotnet/runtimelab/commit/1e3819b16572dd8750947f3979d4cf4b8dfedefa

elinor-fung committed:
Report generator diagnostic when trying to target pre-.NET 5.0 (dotnet/runtimelab#276)



Commit migrated from https://github.com/dotnet/runtimelab/commit/87da2fd706f1e3097c3e471fdf0a5a55758f29bf

elinor-fung committed:
Add analyzer for GeneratedDllImport attribute usage (dotnet/runtimelab#277)



Commit migrated from https://github.com/dotnet/runtimelab/commit/c7a6be69e7c31b932eed4067ecc2afc246bf1040

jkoritzinsky committed:
Array marshalling support (dotnet/runtimelab#272)



Commit migrated from https://github.com/dotnet/runtimelab/commit/bfd398c97a48a66929554af33736057b6fc7b5cf

elinor-fung committed:
Fix byte length calculation for UTF-16 strings (dotnet/runtimelab#285)



Commit migrated from https://github.com/dotnet/runtimelab/commit/a3579c09d520c44eda899847860c5e1a476cbc32

jkoritzinsky committed:
Merge remote-tracking branch 'dotnet/standalone-template' into dllimportgenerator-hookup-build


Commit migrated from https://github.com/dotnet/runtimelab/commit/3e1b2ad5201b0f7e6d76382e7210a774628ee557

elinor-fung committed:
Add string marshallers for ANSI and platform-defined (dotnet/runtimelab#288)

* Update to latest version of analyzer test package
* Add string marshaller for CharSet.Ansi
* Add string marshaller for CharSet.Auto
* Update compatibility doc

Commit migrated from https://github.com/dotnet/runtimelab/commit/ebbd0b2672c13861a2d53a2e7cd8676d67b680e7

elinor-fung committed:
Report diagnostic for `BestFitMapping`, `ThrowOnUnmappableChar`, and `LCIDConversion` (dotnet/runtimelab#291)



Commit migrated from https://github.com/dotnet/runtimelab/commit/633f08d813a2109f390c196465063eef744fe20f

jkoritzinsky committed:
Apply changes to integrate standalone experiment CI/CD pipeline with DllImportGenerator


Commit migrated from https://github.com/dotnet/runtimelab/commit/4956f113781bf92428a7c2676d7a6f97a1a2b8fe

jkoritzinsky committed:
Upgrade DNNE to 1.0.16

Signed-off-by: Jeremy Koritzinsky <jekoritz@microsoft.com>


Commit migrated from https://github.com/dotnet/runtimelab/commit/106540b833576e479efdcd236e73d499dfe9b0b3

jkoritzinsky committed:
Merge pull request dotnet/runtimelab#308 from jkoritzinsky/dllimportgenerator-hookup-build



Commit migrated from https://github.com/dotnet/runtimelab/commit/c6ef6e1625c9ff101f8bfd454c00428436cd2795

elinor-fung committed:
Let enums and pointers go through blittable marshaller (dotnet/runtimelab#303)



Commit migrated from https://github.com/dotnet/runtimelab/commit/7ae2d22408bb1ef598f8a73ad7b7fac904b3e157

jkoritzinsky committed:
Disable packing Demo app, tools, and test assets. (dotnet/runtimelab#327)



Commit migrated from https://github.com/dotnet/runtimelab/commit/93372b1a309f215001f22cdde6b240454585521f

elinor-fung committed:
Fix formatting for some diagnostic messages (dotnet/runtimelab#311)



Commit migrated from https://github.com/dotnet/runtimelab/commit/3e23af10ec81a7715c1d5c151debe920d1e80c9a

elinor-fung committed:
Handle PreserveSig=false (dotnet/runtimelab#323)

* Handle PreserveSig=false
* Flag SetLastError=true as unsupported until we add support for it
* Add try-finally to generated code
* Update doc

Commit migrated from https://github.com/dotnet/runtimelab/commit/4562fa702e3cd56ce293dee960ffdf1bef7c68c8

jkoritzinsky committed:
Implement the non-blittable type marshalling proposal (dotnet/runtimelab#302)



Commit migrated from https://github.com/dotnet/runtimelab/commit/fee212a30a6b4aa468fb28a403c86e622988e3bc

elinor-fung committed:
Add DynamicallyAccessedMembers attribute to CreateSafeHandle API (dotnet/runtimelab#373)



Commit migrated from https://github.com/dotnet/runtimelab/commit/e4bdc07d100804a08b00c5419e259ba2ef7a211f

AaronRobinsonMSFT committed:
Apply SkipLocalsInitAttribute to generated stubs. (dotnet/runtimelab#377)

* Apply SkipLocalsInitAttribute to generated stubs.

Commit migrated from https://github.com/dotnet/runtimelab/commit/7ab9caf38b78fd854fe49ce06f85b383edff59a7

elinor-fung committed:
Fix setting of allocation marker in ANSI string marshaller (dotnet/runtimelab#375)



Commit migrated from https://github.com/dotnet/runtimelab/commit/f775098ae51c22f920d842769985d0a0c6917c85

elinor-fung committed:
Handle SetLastError=true (dotnet/runtimelab#360)



Commit migrated from https://github.com/dotnet/runtimelab/commit/b36b0b8adc4d5c3e7f2927381ea800444f870f96

elinor-fung committed:
Add analyzer to flag converting from DllImport to GeneratedDllImport (dotnet/runtimelab#367)



Commit migrated from https://github.com/dotnet/runtimelab/commit/a7ccdd83e4bda565c9a3870ffe48427c23782902

jkoritzinsky committed:
Reliability improvements for array marshalling (dotnet/runtimelab#384)



Commit migrated from https://github.com/dotnet/runtimelab/commit/5a78d17b2ec1f9364446c016a60ca93a449d7e26

AaronRobinsonMSFT committed:
Update prefix of packages and published assemblies (dotnet/runtimelab#423)

* Prefix all packages and and published assemblies with "Microsoft.Interop.*"

Commit migrated from https://github.com/dotnet/runtimelab/commit/e87c78ea898614339f7f043791f1af07d1abf785

jkoritzinsky committed:
Support arrays of types with simple custom marshalling (dotnet/runtimelab#379)



Commit migrated from https://github.com/dotnet/runtimelab/commit/ada198fa6f8b04e735867b3dcd64e1342dbd13db

jkoritzinsky committed:
Implement support for configurable generator options. (dotnet/runtimelab#458)



Commit migrated from https://github.com/dotnet/runtimelab/commit/9649d4319d518db34d531b8111610343f3573e62

jkoritzinsky committed:
Add handling for [In, Out] attributes. (dotnet/runtimelab#380)



Commit migrated from https://github.com/dotnet/runtimelab/commit/18aa6f1961aa891569e384ef9c12e631ce4939d7

jkoritzinsky committed:
Update Compatibility doc for struct marshalling (dotnet/runtimelab#548)



Commit migrated from https://github.com/dotnet/runtimelab/commit/e73c5de99907e5c362bc5290f3b1e2fb942a8e60

jkoritzinsky committed:
Add Benchmarks project. (dotnet/runtimelab#538)



Commit migrated from https://github.com/dotnet/runtimelab/commit/54321e38193b175b4b62b3e2399d4c50a22af5a6

elinor-fung committed:
Test fill-in (dotnet/runtimelab#535)

- array: out byref, return/out with element marshalling
- bool: in by ref, default marshalling
- char: in by ref
- SafeHandle: out by ref

Commit migrated from https://github.com/dotnet/runtimelab/commit/a072658ef3ab7d46b4adffbc3d7a89bf211bf667

elinor-fung committed:
Handle enums when checking if type is considered blittable (dotnet/runtimelab#563)



Commit migrated from https://github.com/dotnet/runtimelab/commit/c41c1171061920c5cdfa54633cb590bc6417d2e3

elinor-fung committed:
Add fixer for converting to GeneratedDllImport (dotnet/runtimelab#564)



Commit migrated from https://github.com/dotnet/runtimelab/commit/dd810614463b99d14972fe1f4f54f3215212e4e9

jkoritzinsky committed:
Update SafeHandle codegen to match the approved API. (dotnet/runtimelab#570)



Commit migrated from https://github.com/dotnet/runtimelab/commit/0c64a2aa529dff4ef3ad15b17663b5314f55145b

jkoritzinsky committed:
Fix build errors from an updated SDK/compiler and floating package versions. (dotnet/runtimelab#944)

Centralize package versions in Versions.props so we have one point for updating a given package across the whole repo.

Commit migrated from https://github.com/dotnet/runtimelab/commit/9aac13827fc9dcef80dfb44a3dab729a331762ae

jkoritzinsky committed:
Fix cases when preprocessor definitions are surrounding the method we are generating. (dotnet/runtimelab#947)



Commit migrated from https://github.com/dotnet/runtimelab/commit/600ff5af94ad6d1793787e62af75a7bf8ec7142a

elinor-fung committed:
Rename MarshalEx.SetLastWin32Error -> SetLastPInvokeError (dotnet/runtimelab#1007)



Commit migrated from https://github.com/dotnet/runtimelab/commit/82735dd0fef4886ae2f0886267ef30822e17ee8a

elinor-fung committed:
Update to Arcade 6.0.0-beta.21222.1 (dotnet/runtimelab#1008)

* Update to Arcade 6.0.0-beta.21222.1

* Set DNNE roll forward to Major so that we can build with 6.0 SDK

Commit migrated from https://github.com/dotnet/runtimelab/commit/cd7b25b549a8719809d85f9d6292683f1a0a117e

elinor-fung committed:
Stop using myget feed (dotnet/runtimelab#1022)



Commit migrated from https://github.com/dotnet/runtimelab/commit/984f900636a641529213035839fd872833ec1121

jkoritzinsky committed:
Update analysis warning location model to new syntax and clean up analyzer warning locations. (dotnet/runtimelab#984)



Commit migrated from https://github.com/dotnet/runtimelab/commit/88a78946cf8a246348d48e7c42c6666cf57d68c3

elinor-fung committed:
Merge latest standalone-template into feature/DllImportGenerator (dotnet/runtimelab#1031)

* Fix runsettings generation

* Update standalone-template for recent infrastructure changes (dotnet/runtimelab#1024)

* Update to Arcade 6.0.0-beta.21226.16

* Update dependency for using custom runtime version

* Update README

* Don't use custom runtime in Ancillary.Interop

Co-authored-by: Santiago Fernandez Madero <safern@microsoft.com>

Commit migrated from https://github.com/dotnet/runtimelab/commit/57b11ae82fd1f4fbde9a98e415082db6087d0142

elinor-fung committed:
Move docs to top-level directory (dotnet/runtimelab#1034)



Commit migrated from https://github.com/dotnet/runtimelab/commit/4a232c1a38dcb88a2dd6755672e5c419f241eeff

elinor-fung committed:
Add DLLIMPORTGENERATOR_ENABLED when DllImportGenerator package is referenced (dotnet/runtimelab#1050)



Commit migrated from https://github.com/dotnet/runtimelab/commit/e1116095420cecf0d28f04d8e422653761fc37dc

jkoritzinsky committed:
Switch to the now-built-in Marshal APIs from MarshalEx (dotnet/runtimelab#1042)



Commit migrated from https://github.com/dotnet/runtimelab/commit/26dec3e1ca3fac21946bfbcc3031bd2017929c10

elinor-fung committed:
Handle function pointers (dotnet/runtimelab#1043)



Commit migrated from https://github.com/dotnet/runtimelab/commit/a33df5c87196361a4a7d13b7d77e440aae5706f8

kant2002 committed:
Fix link (dotnet/runtimelab#1065)



Commit migrated from https://github.com/dotnet/runtimelab/commit/2fdea5e3a52219e6a5d200a96589a196067b7eb5

jkoritzinsky committed:
Add support for null arrays in the blittable array marshaller's pinning optimization. (dotnet/runtimelab#1063)

* Add support for null arrays in the blittable array marshaller's pinning optimization.

* Add comment block about the behavior.

* Use pointer casting instead of Unsafe.NullRef since we are already using pointers.

Commit migrated from https://github.com/dotnet/runtimelab/commit/c436fda400ce8b6bc374a69f53d02091d8e968f3

jkoritzinsky committed:
Add support for abstract SafeHandle types for by-value marshalling. (dotnet/runtimelab#1066)

* Add support for abstract SafeHandle types for by-value marshalling.

* Add test for byref abstract SafeHandle.

* Add abstract by-ref test. Add details for failure case.

Commit migrated from https://github.com/dotnet/runtimelab/commit/a406e64e53db09de69f6125b2c7d4e2311b7120d

jkoritzinsky committed:
Disable the test for the Marshal switch now that it currently is a no-op. I've disabled the test so that we can re-enable it if we end up adding more APIs to the MarshalEx surface. (dotnet/runtimelab#1082)



Commit migrated from https://github.com/dotnet/runtimelab/commit/b7688d5d93b35256521b2460c5b6fcdc256bd3bb

jkoritzinsky committed:
Keep target DllImport info in structured data before converting to syntax. (dotnet/runtimelab#1075)



Commit migrated from https://github.com/dotnet/runtimelab/commit/2f89e87e872378f5fafe907a6e288d9f1784e55a

jkoritzinsky committed:
Make DllImport target function local. (dotnet/runtimelab#1090)



Commit migrated from https://github.com/dotnet/runtimelab/commit/b8c7d30a556400cc6f5d6f3616863e16086a33a4

elinor-fung committed:
Basic infrastructure to enable DllImportGenerator (#52486)

jkoritzinsky committed:
Prepend identifiers with @ symbol since the Roslyn Symbol API strips the @ from the name. (dotnet/runtimelab#1105)

Fixes https://github.com/dotnet/runtimelab/issues/1096

Commit migrated from https://github.com/dotnet/runtimelab/commit/3c162b8e3a01a3753db5411cd2d67719606868ce

jkoritzinsky committed:
Use SemanticModel to discover GeneratedDllImportAttribute. (dotnet/runtimelab#1116)



Commit migrated from https://github.com/dotnet/runtimelab/commit/4dc8111108263a9f17d9c18d9a55ffbaf767b21f

jkoritzinsky committed:
Support implicit blittability for structs declared in and not exposed outside of the current compilation. (dotnet/runtimelab#1126)



Commit migrated from https://github.com/dotnet/runtimelab/commit/51d9ad880e106c88188630fd9318360b1ac21854

elinor-fung committed:
Move to DllImportGenerator 1.0.0-alpha.21268.4 (#52981)

elinor-fung committed:
Use GeneratedDllImport in System.Diagnostics.FileVersionInfo and System.Runtime.InteropServices.RuntimeInformation (#52739)

* Convert System.Diagnostics.FileVersionInfo
* Convert System.Runtime.InteropServices.RuntimeInformation
* Use SYSTEM_INFO* instead of GeneratedDllImport for GetSystemInfo/GetNativeSystemInfo

elinor-fung committed:
Use GeneratedDllImport in System.Console (#52740)

jkoritzinsky committed:
Require that methods marked with GeneratedDllImport are in types that are marked partial and any parents of their containing types are also marked partial. (dotnet/runtimelab#1091)



Commit migrated from https://github.com/dotnet/runtimelab/commit/53af4b5dfa7e4d77b34151523e82665974da329c

elinor-fung committed:
Merge remote-tracking branch 'upstream/main' into feature/use-dllimport-generator

elinor-fung committed:
Use GeneratedDllImport in System.Security.Cryptography.Encoding (#53111)

elinor-fung committed:
Use GeneratedDllImport in System.Security.Cryptography.Algorithms (#53245)

elinor-fung committed:
Use GeneratedDllImport in System.Security.Cryptography.X509Certificates (#53259)

jkoritzinsky committed:
Introduce design doc for Span/ReadOnlySpan<T> marshalling (dotnet/runtimelab#1103)

* First draft of the heart of the span marshalling design.

Finish the in-code collection marshalling support design doc.

* Apply suggestions from code review

Co-authored-by: Aaron Robinson <arobins@microsoft.com>

* Feedback.

* Empty spans will be marshalled as null.

* Make the primary design in Design 2 be contiguous collection support.

* pr feedback and add example stub.

* Fix typo

Co-authored-by: Aaron Robinson <arobins@microsoft.com>

Commit migrated from https://github.com/dotnet/runtimelab/commit/1bc504d822114b70c8a3df5709b39484bc99a911

AaronRobinsonMSFT committed:
Handle case where void* is a field. (dotnet/runtimelab#1195)

* Handle case where void* is a field.

Commit migrated from https://github.com/dotnet/runtimelab/commit/41bf5d17bc65e44fc8f8bba118ddb5c51b77ed1e

jkoritzinsky committed:
Fix memory leaks for return values and some out params. (dotnet/runtimelab#1200)

Fix some memory leaks we had where we didn't generate code to free native resources for out parameters that use custom native type marshalling or all return values.

Also set a flag in the csproj that enables a nice debugging innerloop option for roslyn components in the csproj settings (in our case running the generator on the integration tests)

Commit migrated from https://github.com/dotnet/runtimelab/commit/b11a26add0e6a599d3f87d73015988034c665583

elinor-fung committed:
Use GeneratedDllImport in System.Diagnostics.Process (#53702)

AaronRobinsonMSFT committed:
Use GeneratedDllImport in System.IO.FileSystem (#53830)

* Convert System.IO.FileSystem to DllImport source generator

elinor-fung committed:
Use GeneratedDllImport in System.Net.Security (#53839)

AaronRobinsonMSFT committed:
Comment in source that source file is auto-generated. (dotnet/runtimelab#1209)

* Comment in source that source file is auto-generated.

Co-authored-by: Sam Harwell <sam@tunnelvisionlabs.com>

Commit migrated from https://github.com/dotnet/runtimelab/commit/fd4042ce8f0f90d82e34f70cec03056db14717cb

elinor-fung committed:
Use GeneratedDllImport in System.IO.FileSystem.AccessControl (#53845)

elinor-fung committed:
Use GeneratedDllImport in System.Net.Sockets (#53844)

elinor-fung committed:
Use GeneratedDllImport in System.Security.Principal.Windows (#53903)

elinor-fung committed:
Use GeneratedDllImport in System.Net.Http (#53885)

elinor-fung committed:
Use GeneratedDllImport in System.Net.HttpListener (#53904)

jkoritzinsky committed:
Implement collection marshaller spec (dotnet/runtimelab#1197)



Commit migrated from https://github.com/dotnet/runtimelab/commit/b3a0ead76a03a92af3b87dd5486563728fb5ecf8

elinor-fung committed:
Use GeneratedDllImport in System.IO.Pipes (#53947)

jkoritzinsky committed:
Add configuration option for Unsafe type name. (dotnet/runtimelab#1217)



Commit migrated from https://github.com/dotnet/runtimelab/commit/5fa8059e0ba71465be018d97a37db93d72bbcb9e

elinor-fung committed:
Don't enable DllImport generator for shims (#53973)

AaronRobinsonMSFT committed:
Use GeneratedDllImport in System.Net.NetworkInformation (#53966)

* Use GeneratedDllImport in System.Net.NetworkInformation

elinor-fung committed:
Use GeneratedDllImport in System.IO.MemoryMappedFiles (#53978)

elinor-fung committed:
Use DLLIMPORTGENERATOR_ENABLED ifdefs for files shared with Microsoft.IO.Redist (#53954)

jkoritzinsky committed:
Enable marshalling collections of types with marshallers that have a Value property (dotnet/runtimelab#1216)



Commit migrated from https://github.com/dotnet/runtimelab/commit/c1562420e508e4939f138cc9d5145a8c43fa7afc

jkoritzinsky committed:
Handle native identifiers for escaped managed identifiers. (dotnet/runtimelab#1167)



Commit migrated from https://github.com/dotnet/runtimelab/commit/4835f732a212ebe2ffc06760d9925cbdfec10a11

elinor-fung committed:
Merge remote-tracking branch 'upstream/main' into HEAD

elinor-fung committed:
Make HmacOneShot blittable

jkoritzinsky committed:
Implement support for collections of collections marshalling (dotnet/runtimelab#1226)

* Refactor indexer naming to be nested-collection safe.

* Implement initial collections-of-collections/jagged array support.

Use a topological sort of the parameter+return marshallers to unmarshal "CountElementName"-referenced parameters/return before unmarshalling the elements that have a dependency on them through "CountElementName".

* Fixes for edge cases around HResult/Exception handling.

* Flip the edgeMap indices so we can use Array.IndexOf (which is optimized) to search for edges.

* Comments and optimizations.

* Add citation for algorithm,

* Hoist elementIndex out of the loop.

* Encapsulate the edgeMap in a private struct type and add some simple abstractions to enable more perf optimizations.

* Make Topological sort more flexible for element ids

* Use native index if managed index is unset to allow handling multiple native-only parameters as distinct nodes in the graph model.

* Fix using the return value for passing collection size. Validate the types of the CountElementName'd elements, even in nested scenarios.

* Change how we initialize the numRows array.

* Fix cycle breaking in count info.

* Update nested indexer creation for num elements expressions to handle non-collection sub contexts.

* Add stress test for collections of collections that uses 0-11 nested arrays.

* PR feedback.

Commit migrated from https://github.com/dotnet/runtimelab/commit/aa2c08cb60f4e13e06e50be2d8c6e7f50b6ec5da

jkoritzinsky committed:
Implement marshallers for Span and ReadOnlySpan (dotnet/runtimelab#1222)

* Implement marshallers for Span/ReadOnlySpan.

* Add tests for span marshalling.

* Remove unused code.

* PR feedback.

* Use dotnet/runtime style.

* Fix allocation size in ReadOnlySpanMarshaller<T>.

Commit migrated from https://github.com/dotnet/runtimelab/commit/e9876f36bda716892f6a3878b896991e1be02b81

elinor-fung committed:
Use GeneratedDllImport in System.IO.Compression.Brotli (#54173)

elinor-fung committed:
Use GeneratedDllImport in System.IO.FileSystem.DriveInfo (#54181)

elinor-fung committed:
Use GeneratedDllImport in System.IO.FileSystem.Watcher (#54183)

elinor-fung committed:
Use GeneratedDllImport in System.Net.NameResolution (#54191)

kant2002 committed:
Fix table display (dotnet/runtimelab#1242)



Commit migrated from https://github.com/dotnet/runtimelab/commit/35df9046707caf1cb84f476cc443e5828ce214f7

elinor-fung committed:
Use GeneratedDllImport in System.Net.Ping (#54232)

AaronRobinsonMSFT committed:
Remove fixed workarounds (dotnet/runtimelab#1247)



Commit migrated from https://github.com/dotnet/runtimelab/commit/dbc21f4fd959331784f68da10c3772adf06895a6

AaronRobinsonMSFT committed:
Use GeneratedDllImport in Microsoft.Win32.Registry (#54236)

elinor-fung committed:
Use GeneratedDllImport in System.Security.Cryptography.Csp (#54290)

elinor-fung committed:
Merge remote-tracking branch 'upstream/main' into mergeMain

jkoritzinsky committed:
Forward special P/Invoke attributes to the target DllImport method (dotnet/runtimelab#1267)



Commit migrated from https://github.com/dotnet/runtimelab/commit/5715e2ee4fd7c1c81151004f6332bf66731e537f

jkoritzinsky committed:
Fix SafeHandle benchmarks and add string benchmarks (dotnet/runtimelab#1238)



Commit migrated from https://github.com/dotnet/runtimelab/commit/165fe8cfacc9c46df8062a9d8f975c35aa52cd89

jkoritzinsky committed:
Only run GuaranteedUnmarshal statements if invoke was successful. (dotnet/runtimelab#1288)



Commit migrated from https://github.com/dotnet/runtimelab/commit/de39034ce26e5ab47c0967748bc839bd40e58266

jkoritzinsky committed:
Don't emit SkipLocalsInit when it's already active (dotnet/runtimelab#1289)

* Don't emit SkipLocalsInit on each stub if it's already specified on the module as a whole, the containing type, or the stub method

* Precalculate module check and check that first.

Commit migrated from https://github.com/dotnet/runtimelab/commit/7442710c450a3799e6d60bd57604741f88f18b73

jkoritzinsky committed:
Require using UnmanagedCallConv instead of the CallingConvention field on GeneratedDllImport (dotnet/runtimelab#1292)

Co-authored-by: Aaron Robinson <arobins@microsoft.com>

Commit migrated from https://github.com/dotnet/runtimelab/commit/f22931b97e89714a482307c160421eca70ad54f6

jkoritzinsky committed:
Report diagnostics for bad opt-in marshallers at MarshalUsing use-site if the marshaler is in a different compilation than the current compilation. (dotnet/runtimelab#1300)



Commit migrated from https://github.com/dotnet/runtimelab/commit/d4b5fd03f51ec7f932f3e8b5856b8233903e9a39

jkoritzinsky committed:
Resolve SizeParamIndex to a TypePositionInfo during MarshallingInfo parsing (dotnet/runtimelab#1293)



Commit migrated from https://github.com/dotnet/runtimelab/commit/eea0baeaa690b74e4d2d443d5cf951444b380e06

jkoritzinsky committed:
Update analyzer to correctly handle collection marshallers (dotnet/runtimelab#1299)



Commit migrated from https://github.com/dotnet/runtimelab/commit/c11f860f6f5935ac6d2db35ea64f20abb8f562c6

jkoritzinsky committed:
Improve diagnostics messages for invalid custom native type marshalling and cyclical element info. (dotnet/runtimelab#1306)

Co-authored-by: Elinor Fung <elfung@microsoft.com>

Commit migrated from https://github.com/dotnet/runtimelab/commit/cd18bc47d5be7f18e80aef8c22f8848bed8cdffe

jkoritzinsky committed:
Merge branch 'main' of github.com:dotnet/runtime into feature/use-dllimport-generator

jkoritzinsky committed:
Merge pull request #56565 from jkoritzinsky/merge-main

elinor-fung committed:
Update compat doc (dotnet/runtimelab#1379)



Commit migrated from https://github.com/dotnet/runtimelab/commit/1cf4425379bfb119d2eed24826bf21aa7ab70a3e

jkoritzinsky committed:
Use the repo-local NuGet.config with GetReferenceAssemblies so we can use pre-release ref packs reliably. (dotnet/runtimelab#1375)



Commit migrated from https://github.com/dotnet/runtimelab/commit/397ccce0d8612c3655878ea83466941276b32c43

AaronRobinsonMSFT committed:
Plan for .NET 7 integration (dotnet/runtimelab#1378)

* Plan for .NET 7 integration

Commit migrated from https://github.com/dotnet/runtimelab/commit/cd76a36d7d7744bd52c65b9f5ff3d303900cb687

jkoritzinsky committed:
Move to using the new Roslyn IIncrementalGenerator API for better in-VS performance (dotnet/runtimelab#1374)



Commit migrated from https://github.com/dotnet/runtimelab/commit/90c617d39a9775e8eb0c5d4aba2f0cc6932bfb87

jkoritzinsky committed:
Refactor DllImportGenerator project for easier extensibility (dotnet/runtimelab#1119)



Commit migrated from https://github.com/dotnet/runtimelab/commit/9ac16c6c9f47100ea91db6231ad47f4a8d6ca946

AaronRobinsonMSFT committed:
Remove duplicate TypeNames.cs and create IntPtr constant. (dotnet/runtimelab#1550)



Commit migrated from https://github.com/dotnet/runtimelab/commit/fb867cdfb646a4eecddb3eb91baa4709811872b5

jkoritzinsky committed:
Update DllImportGenerator to the latest version (#59117)

jkoritzinsky committed:
Merge branch 'main' of https://github.com/dotnet/runtime into merge-main

jkoritzinsky committed:
Fix failures in the AllConfigurations build.

jkoritzinsky committed:
Merge pull request #59119 from jkoritzinsky/merge-main

jkoritzinsky committed:
Merge branch 'feature/DllImportGenerator' of /home/jekoritz/scratch/bash/dllimportgenerator into port-dllimport-generator

jkoritzinsky committed:
Hook up DllImportGenerator to the libraries build (excluding solution file regeneration).

Move System.Runtime.InteropServices unit tests to a subdirectory.

jkoritzinsky committed:
Run slngen to regenerate the solution files.

jkoritzinsky committed:
Merge pull request #59363 from jkoritzinsky/port-dllimport-generator

Port DllImportGenerator to dotnet/runtime

jkoritzinsky committed:
Merge branch 'main' of github.com:dotnet/runtime into merge-main

jkoritzinsky committed:
Merge pull request #59577 from jkoritzinsky/merge-main

am11 commented:
Do we want to keep the new `src/samples/` directory, or move it to https://github.com/dotnet/samples?

jkoritzinsky commented:
The new src/samples directory is partially meant for quick and simple testing with native debuggers without having to go through `dotnet test`. Very similar to AndroidSampleApp and the other samples in src/mono/samples/.

jkotas commented:
> The new src/samples directory is partially meant for quick and simple testing with native debuggers 

Should it have more appropriate name (e.g. `TestApps`) to capture the intent and make it obvious that these are not meant to be user-facing samples?

Also, should it have less stuff in it (e.g. just a single signature)?

jkoritzinsky committed:
Remove target framework in the call to DllImportGenerator.csproj.

This fixes an issue where a project that targets net7.0-windows accidentally ends up getting an output path pointing to the wrong directory for Microsoft.Interop.SourceGeneration.dll, which causes us to hit https://github.com/dotnet/roslyn/issues/56442.

AaronRobinsonMSFT committed:
Add license banner to source. (#59586)

elinor-fung committed:
Enable running analyzers on DllImportGenerator and Microsoft.Interop.SourceGeneration (#59683)

elinor-fung committed:
Make Utf16CharMarshaller pin ref parameters (#59700)

elinor-fung committed:
Fix null termination for generated marshalling of read-only ref string (#59726)

jkoritzinsky committed:
Merge branch 'main' into feature/use-dllimport-generator

elinor-fung committed:
Update DllImportGenerator style to match repo conventions (#59753)

elinor-fung committed:
Change some P/Invokes back to non-blittable + GeneratedDllImport for better usability (#59876)

jkoritzinsky committed:
Merge branch 'main' into feature/use-dllimport-generator

jkoritzinsky committed:
Merge branch 'main' into feature/use-dllimport-generator

elinor-fung committed:
Fix IDE0008 in DllImportGenerator and Microsoft.Interop.SourceGeneration (#59885)

elinor-fung committed:
Fix inconsistencies in order of GeneratedDllImport fields (#60049)

jkoritzinsky committed:
Dllimport generator build and test fixes (#59658)

am11 commented on src/libraries/System.Runtime.InteropServices/tests/TestAssets/NativeExports/NativeExports.csproj:
```diff
@@ -17,4 +20,97 @@
     <ProjectReference Include="..\SharedTypes\SharedTypes.csproj" />
   </ItemGroup>
 
+  <Target Name="GetLinuxCrossBuildArgumentsForDNNE"
```
Can the DNNE build moved under `src/tests`where common native infra is already wired up? We can add a simple cmake file for DNNE project there (as we have for other external projects  libunwind, brotli etc.), without replicating the script code so implicitly.

jkoritzinsky replied:
We'd rather not move the DNNE project stuff over to the `src/tests` tree as that will prevent us from using XUnit for our tests easily. We're also trying to keep the DNNE native component very slim and as transparent to the build as possible, and moving it into the CMake build for tests would make it not transparent due to build dependencies (the native C file is generated from the compiled managed assembly)

jkoritzinsky committed:
Fix incorrect endif location that broke the build.

jkoritzinsky committed:
Merge branch 'main' into feature/use-dllimport-generator

jkoritzinsky commented:
I've run the ASP.NET JSON macrobenchmark on this branch and compared it to main, and we see a minor performance improvement in steady state performance with this branch.

<summary>Baseline</summary>
<details>

| application           |                                |
| --------------------- | ------------------------------ |        
| CPU Usage (%)         | 97                             |        
| Cores usage (%)       | 1,166                          |        
| Working Set (MB)      | 182                            |        
| Private Memory (MB)   | 771                            |        
| Build Time (ms)       | 13,179                         |
| Start Time (ms)       | 184                            |        
| Published Size (KB)   | 109,864                        |        
| .NET Core SDK Version | 5.0.401                        |        
| ASP.NET Core Version  | 5.0.10+b92c9f5                 |        
| .NET Runtime Version  | 7.0.0-alpha.1.21507.12+8fdd22d |        


| load                   |                |
| ---------------------- | -------------- |
| CPU Usage (%)          | 62             |
| Cores usage (%)        | 745            |
| Working Set (MB)       | 50             |
| Private Memory (MB)    | 358            |
| Build Time (ms)        | 7,140          |
| Start Time (ms)        | 0              |
| Published Size (KB)    | 76,412         |
| .NET Core SDK Version  | 3.1.413        |
| ASP.NET Core Version   | 3.1.19+b908e91 |
| .NET Runtime Version   | 3.1.19+aae0024 |
| First Request (ms)     | 123            |
| Requests/sec           | 469,558        |
| Requests               | 7,090,260      |
| Mean latency (ms)      | 0.79           |
| Max latency (ms)       | 48.15          |
| Bad responses          | 0              |
| Socket errors          | 0              |
| Read throughput (MB/s) | 68.07          |
| Latency 50th (ms)      | 0.50           |
| Latency 75th (ms)      | 0.58           |
| Latency 90th (ms)      | 0.72           |
| Latency 99th (ms)      | 10.98          |

</details>

<summary>This PR</summary>
<details>

| application           |                                |
| --------------------- | ------------------------------ |        
| CPU Usage (%)         | 97                             |        
| Cores usage (%)       | 1,168                          |        
| Working Set (MB)      | 185                            |        
| Private Memory (MB)   | 733                            |        
| Build Time (ms)       | 10,301                         |        
| Start Time (ms)       | 197                            |
| Published Size (KB)   | 110,366                        |        
| .NET Core SDK Version | 5.0.401                        |        
| ASP.NET Core Version  | 5.0.10+b92c9f5                 |        
| .NET Runtime Version  | 7.0.0-alpha.1.21507.14+78f4fa8 |        


| load                   |           |
| ---------------------- | --------- |
| CPU Usage (%)          | 63        |
| Cores usage (%)        | 754       |
| Working Set (MB)       | 38        |
| Private Memory (MB)    | 358       |
| Start Time (ms)        | 0         |
| First Request (ms)     | 129       |
| Requests/sec           | 471,137   |
| Requests               | 7,114,106 |
| Mean latency (ms)      | 0.78      |
| Max latency (ms)       | 52.35     |
| Bad responses          | 0         |
| Socket errors          | 0         |
| Read throughput (MB/s) | 68.30     |
| Latency 50th (ms)      | 0.50      |
| Latency 75th (ms)      | 0.58      |
| Latency 90th (ms)      | 0.71      |
| Latency 99th (ms)      | 10.73     |
</details>

elinor-fung committed:
Updates to DllImportGenerator docs (#60211)

jkoritzinsky committed:
Re-enable running DllImportGenerator.Tests suite on Mono. (#60259)

AaronRobinsonMSFT commented:
> and there is no significant gain in performance either.

@kasperk81 This is a short term and intentional non-goal. The reasoning is we want to ensure a high level of confidence and compatibility for v1. To that end we did do some optimizations, but very few. As we gain confidence with how this methodology works we are free to try optimizations that are targeted as opposed to the "everyone gets it" design with the built-in system. As we get further into .NET 7 and onto .NET 8 we are considering opportunities for better performance in this area.

jkoritzinsky committed:
Disable ClearPreviousError test on Mono (#60272)

AaronRobinsonMSFT committed:
Add design details regarding dynamic buffers in struct marshalling (#60374)

* Update StructMarshalling.md with details

Co-authored-by: Elinor Fung <elfung@microsoft.com>
Co-authored-by: Jeremy Koritzinsky <jekoritz@microsoft.com>

elinor-fung committed:
Undo solution and config changes from DllImportGenerator (#60481)

* Remove dotnet-experimental feed from NuGet.config
* Revert changes to solution files and restore configs

elinor-fung committed:
Merge branch 'main' into feature/use-dllimport-generator

AaronRobinsonMSFT reviewed (approved):
LGTM
